### ISSUE TITLE
[2017 ICML] A Closer Look at Memorization in Deep Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#72]
Neural net이 dataset을 memorize할 수 있음을 보여준 `Understanding deep learning requires rethinking generalization` 논문을 발전시킨 논문

저자들은 real data와 noise data간의 DNN의 learning dynamics를 실험하며 main contribution은 다음과 같음

1. Read data와 noise data에서 DNN optimization behavior에 qualitative difference가 존재
2. DNN은 memorize하기 전에 multiple training examples간에 공유되는 simple pattern을 먼저 학습함 (DNN optimization is content-aware)
3. Regularization technique를 통해 DNN이 real data에서 학습할 수 있는 능력을 보존하면서 memorization을 막을 수 있음

편의상 randX data를 input image가 gaussian noise인 경우, randY data는 image에 random label이 붙은 경우로 명칭

### 1.1. Easy Examples as Evidence of Patterns in Real Data
<img width="552" alt="image" src="https://user-images.githubusercontent.com/49643709/170838556-9f266533-5498-455d-abae-b44eb732bceb.png">
<img width="546" alt="image" src="https://user-images.githubusercontent.com/49643709/170838588-9f3bd66c-67a0-4ac0-891f-800682ccacf3.png">

만약 network가 data의 pattern에 기반하여 학습한다면, real data에서 몇몇 이미지들은 다른 이미지들보다 더 학습하기 쉬울 것임

즉 network가 어느정도 학습했을 때, 학습하기 쉬운 easy examples에 대해서는 일관되게 잘 맞출 것이고 학습하기 어려운 hard examples에 대해서는 일관되게 못 맞출 것임

RandX data는 examples들이 학습의 difficulty 차이는 거의 없기에, network가 어느정도 학습했을 때, 일관되지 않은 결과를 보일 것임

실험 결과를 통해 real data의 example별 difficulty가 다양함을 확인할 수 있음

Binomial distribution sampling을 통해 randX data의 example별 difficulty는 거의 동일함을 보임

RandY data는 example별 difficulty가 어느정도 있음을 확인할 수 있으나, real data와 다른 것을 배움을 확인할 수 있음
(conv1 visualization 확인)

#### 1.2.1. Loss-sensitivity
<img width="878" alt="image" src="https://user-images.githubusercontent.com/49643709/170838889-1c076213-796a-41ad-a29e-8062e4ac7143.png">
<img width="560" alt="image" src="https://user-images.githubusercontent.com/49643709/170838966-51b97996-1811-453f-8080-c71018cc083f.png">

저자들은 input이 loss에 얼마나 영향을 주는지를 measure하는 loss-sensitivity를 제안함

Loss-sensitivity가 크다는 말은 input이 조금만 바뀌어도 loss가 많이 변화한다는 뜻임

Real data에서는 몇몇 data만 loss-sensitivity가 큰데, randX data에서는 대부분의 data가 큰 loss-sensitivity를 가짐

이를 통해 neural network는 아마도 memorize하지 않을 것이며, 적어도 randX data와는 다른 학습 양상을 보인다는 것을 보임

참고로 gini coefficient는 value의 inequality를 measure하며, 0은 exact equality를 1은 maximal inequality를 의미함

#### 1.2.2. Class specific loss-sensitivity
<img width="570" alt="image" src="https://user-images.githubusercontent.com/49643709/170838909-3c2dd5f2-a428-4870-9c8b-e2bcab665756.png">
<img width="553" alt="image" src="https://user-images.githubusercontent.com/49643709/170838874-27513046-820a-48ae-aa52-ab383cf9555c.png">

저자들은 class specific loss-sensitivity를 제안하며 실험을 진행

Real data는 loss-sensitivity가 spread되어있으며 높은데, 이는 cross-category pattern으로 인해 발생하는 것으로 볼 수 있음

예를 들면 개, 고양이 사이의 cross-category pattern으로 인해, 개 이미지에 대한 고양이 label의 loss-sensivity가 높다는 것임

RandX data는 당연히 cross-category pattern이 없기에 diagonal한 term만 높음

#### 1.3.1. Effects of capacity and data size on validation performances
<img width="554" alt="image" src="https://user-images.githubusercontent.com/49643709/170839706-a1634200-74dd-4080-8731-19a6390760d5.png">

다양한 level의 randX data에서 model capacity와 validation performance의 관계를 알기 위해 실험
(level이 높을수록 randX의 비율이 더 많아짐)

RandX level이 높을수록 optimal validation performance를 위해 higher capacity model을 사용해야함

해당 결과는 most regular patterns를 배우기 위해 capacity를 제한해야한다는 traditional learning theory와 반대되는 결과임

저자들은 network가 higher capacity를 가지면 real data의 학습을 방해하지 않는 방식으로 randX examples를 학습할 수 있다고 hypothesize

RandY data에서는 capacity와 performance간의 관계를 찾을 수 없었음

#### 1.3.2. Effects of capacity and dataset size on training time
<img width="1133" alt="image" src="https://user-images.githubusercontent.com/49643709/170840155-05ed4996-be7f-4fd5-bca6-699d64695420.png">

다양한 level의 randX data에서 model의 capacity가 줄거나 dataset size가 늘어났을 때, time-to-convergence과의 관계를 알기 위해 실험

Neural network의 capacity가 커지면 time-to-convergence가 작아지는데, randX level이 클수록 더 영향을 많이 받음

Dataset size가 늘어나면 time-to-convergence가 커짐

Input과 output간의 관계가 유의미해지면, input to output mapping을 하는데 있어서 더 많은 clues를 주기에 학습시간이 줄어들 것임

Real data에서 dataset size가 50000일 때가 180일 때보다 학습시간이 줄어들기에, network가 memorizing을 한다기보다는 data에서의 pattern을 학습한다는 것을 알 수 있음
(data가 작을 때에는 pattern이 잘 안보여서 clue가 없었을 것임)

### 2.1. Critical Sample Ratio (CSR)
<img width="553" alt="image" src="https://user-images.githubusercontent.com/49643709/170840670-20a2db08-077a-41f4-b532-4a4cd274eed4.png">
<img width="549" alt="image" src="https://user-images.githubusercontent.com/49643709/170840680-8e4c6a8c-a6cd-4c65-aa27-7884afa7eb1e.png">

저자들은 LASS 알고리즘을 통해 찾은 adversarial sample들의 개수를 이용하여 decision boundary의 density를 measure하는 critical sample ratio를 제안함

### 2.2. Critical Samples Throughout Training
<img width="558" alt="image" src="https://user-images.githubusercontent.com/49643709/170840838-aba8acf9-7b24-42a8-a182-82785d795755.png">

RandX data, randY data에서 학습한 neural network의 critical samples가 더 많음

이 말은 randX data, randY data에서 학습한 decision surface가 real data보다 더 complex하다는 것을 의미함

또한 학습이 진행됨에 따라 CSR이 증가하다 stabilize하는데, 이는 network가 점진적으로 더 complex hypotheses를 배운다는 것을 알 수 있음

<img width="1096" alt="image" src="https://user-images.githubusercontent.com/49643709/170840987-bcb76890-6dbd-4ca9-8b7b-da17538f89b2.png">
<img width="1081" alt="image" src="https://user-images.githubusercontent.com/49643709/170840994-baf7bdaf-ea34-4a23-a20b-690a6a465689.png">

Dataset이 noisy할수록 CSR이 높음

RandY dataset에서 training set에서 high accuracy를 도달하기 전에 maximum validation set accuracy를 가짐

이 말은 모델은 먼저 real data의 simple, general pattern을 먼저 학습한 후 randY data를 fitting한다는 것을 알 수 있음

## 3. Effect of Regularization on Learning
<img width="572" alt="image" src="https://user-images.githubusercontent.com/49643709/170841307-a26196a2-d7c5-4fef-961a-475f257afbd6.png">
<img width="561" alt="image" src="https://user-images.githubusercontent.com/49643709/170841298-c2e34ce6-dce9-491a-848b-9742c84e96b1.png">

Real data와 randY data가 섞인 dataset에서 explicit regularization을 이용하여 학습을 진행하였으며, explicit regularization을 통해 real data는 정상적으로 학습하면서 noise memorization speed를 줄일 수 있었음

Dropout regularization이 가장 효과적이었음
